ratis-netty/src/main/java/org/apache/ratis/netty/client/NettyClientStreamRpc.java

```diff
@@ -78,7 +80,33 @@ public class NettyClientStreamRpc implements DataStreamClientRpc {
   public static final Logger LOG = LoggerFactory.getLogger(NettyClientStreamRpc.class);
 
   private static class WorkerGroupGetter implements Supplier<EventLoopGroup> {
-    private static final AtomicReference<EventLoopGroup> SHARED_WORKER_GROUP = new AtomicReference<>();
+
+    private static final AtomicReference<ReferenceCountedObject<EventLoopGroup>> SHARED_WORKER_GROUP =
+        new AtomicReference<>();
```

Comment:
We need 
```java
    private static final AtomicReference<ReferenceCountedObject<Supplier<EventLoopGroup>>> SHARED_WORKER_GROUP
        = new AtomicReference<>();
```
since `AtomicReference.updateAndGet(updateFunction)` may call the `updateFunction` without using the result.  The current code will then called `newWorkerGroup(..)` without shutting it down.

Use a `MemoizedSupplier` inside:
```java
        final Supplier<ReferenceCountedObject<Supplier<EventLoopGroup>>> supplier = MemoizedSupplier.valueOf(
            () -> ReferenceCountedObject.wrap(MemoizedSupplier.valueOf(() -> newWorkerGroup(properties))));
        final ReferenceCountedObject<Supplier<EventLoopGroup>> sharedWorkerGroup = SHARED_WORKER_GROUP.updateAndGet(
            g -> g != null ? g : supplier.get());
        return new WorkerGroupGetter(sharedWorkerGroup.retain().get()) {
          ...
        };
```

Reply:
> since AtomicReference.updateAndGet(updateFunction) may call the updateFunction without using the result. The current code will then called newWorkerGroup(..) without shutting it down.

Do you mean it's because `AtomicReference.updateAndGet(updateFunction)` might get re-applied, as mentioned in the `AtomicReference` documentation?

> Since it may be re-applied when attempted updates fail due to contention among threads.

Reply:
Yes.  When there are two or more threads changing an `AtomicReference` at the same time, the `updateAndGet(..)` method runs in a loop and may call the update function multiple times.  In our previous code, it may create multiple `EventLoopGroup` objects and the old objects will not be shutdown.

Reply:
Thank you for the explanation.